### PR TITLE
import heimdall when --instrument is used

### PR DESCRIPTION
**Pull request by runspired**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,6 +1,80 @@
...
-  name: 'ember-heimdall'
+  name: 'ember-heimdall',
+
+  _isInstrumented: null,
+  isInstrumented: function() {
+    if (this._isInstrumented !== null) {
+      return this._isInstrumented;
+    }
+
+    let INSTRUMENT_HEIMDALL = false;
```
Comment (rwjblue) [inline index.js]: Might be simpler to do:

    let INSTRUMENT_HEIMDALL = process.argv.indexOf('--instrument') > -1;

File: index.js [hunk 2]
```diff
@@ -1,6 +1,80 @@
...
+
+    for (let i = 0; i < args.length; i++) {
+      if (args[i] === '--instrument') {
+        INSTRUMENT_HEIMDALL = true;
+        break;
+      }
+    }
+
+    if (INSTRUMENT_HEIMDALL) {
+      this.ui.writeln(chalk.yellow(
```
Comment (runspired) [inline index.js]: apparently code-complete lied to me and this is not the magical incantation

File: index.js [hunk 3]
```diff
@@ -1,6 +1,80 @@
...
+          'Heimdall.js will be included in your build output and instrumentation will not be stripped.'
+        ));
+    }
+
+    this._isInstrumented = INSTRUMENT_HEIMDALL;
+    return INSTRUMENT_HEIMDALL;
+  },
+
+  includedCommands: function() {
+    patchCommand(serveCommand, { name: 'instrument', type: Boolean, default: false });
```
Comment (runspired) [inline index.js]: I'm super not sure this is the best way to do this, I just didn't see another way that also satisfied the constraint of not breaking things that extend the serve/build/test commands

**Comment by runspired** [comment]
cc @twokul

File: index.js [hunk 4]
```diff
@@ -1,6 +1,80 @@
...
+
+// We import these to merge the instrumentation options into the base classes
+// we do this instead of overriding the core commands because other high profile
+// addons such as ember-cordova, ember-cli-nwjs, and ember-electron have had to override
+const serveCommand = require('ember-cli/lib/commands/serve');
+const buildCommand = require('ember-cli/lib/commands/build');
+const testCommand = require('ember-cli/lib/commands/test');
+
+function patchCommand(command, newOption) {
+  command.prototype.availableOptions.push(newOption);
```
Comment (rwjblue) [inline index.js]: We should check for it before pushing, that allows this to become a polyfill if we ever make this an upstream option.

**Review by rwjblue — COMMENTED** [review]
We should implement `cacheKeyForTree` as mentioned in the addon tree cache RFC. Otherwise, we do (a tiny amount) extra work here in `treeForVendor`.

File: index.js (hunk 2 above)
Comment (rwjblue) [reply index.js]: this.ui.writeLine

File: index.js (hunk 3 above)
Comment (twokul) [reply index.js]: you're copying `{ name: 'instrument', type: Boolean, default: false }` 3 times, perhaps a variable?

File: index.js (hunk 4 above)
Comment (runspired) [reply index.js]: 👍

File: tests/dummy/app/templates/application.hbs [hunk 5]
```diff
@@ -1,5 +1,2 @@
-{{!-- The following component displays Ember's default welcome message. --}}
-{{welcome-page}}
-{{!-- Feel free to remove this! --}}
-
-{{outlet}}
+<h1>Application Template</h1>
+{{outlet}}
```
Comment (twokul) [inline tests/dummy/app/templates/application.hbs]: no newline

**Comment by runspired** [comment]
@rwjblue I remember you linked me to that RFC but now I cant find the link

**Comment by runspired** [comment]
nvm found it: https://github.com/ember-cli/rfcs/blob/master/active/0090-addon-tree-caching.md

File: index.js (hunk 3 above)
Comment (runspired) [reply index.js]: resolved

**Pull request merged**